text/0000-partial_types.md

```diff
@@ -0,0 +1,395 @@
+
+Automatically Type-checker gives a mathematical guarantee, because all denied/forbidden fields remain intact! 
+
+And this mean, that Proxy Borrowing borrowing is fully **safe** and _zero cost_ in binary.
+
+## Proxy Borrowing
+
+Borrowing rules for partial types:
+
+`PermittedField` field borrowing rules are ordinary Rust rules. New variable borrows the whole variable (with partial type), but checker pretends it borrows just permitted fields of this variable.
+
+Not-`PermittedField` filed is always is ready to borrow regardless if origin field is denied(by move, by reference, by borrow).
+
+When we write a code for full or partial borrow, the link of object itself returns, but borrow-checker checks to borrow of permitted fields only.
+
+This new mechanism of is simple and universal.
+
+```rust
+struct S4 {a : i32, b : i32, c : i32, d : i32}
+let s = S4 {a : 5, b: 6, c: 7, d: 8};
+    // s : S4
+
+let r_sd = & s.{d};
+    // r_sd : & S4.{d}
+    //
+    // borrow-checker check just for &s.d
+
+let mut mr_sabc = &mut s.{a, b, c};
+    // mr_sabc : &mut S4.{a, b, c}
+    //
+    // borrow-checkercheck just for &mut s.a, &mut s.b, &mut s.c
+
+let rr_sbc = & mr_sabc.{b, c};
+    // rr_sbc : && S4.{b, c}
+    //
+    // borrow-checker check just for &mr_sabc.b, &mr_sabc.c
+
+let mut mrr_sa = &mut mr_sabc.{a};
+    // mrr_sa : &&mut S4.{a}
+    //
+    // borrow-checker check just for &mut mr_sabc.a
+```
+
+## Syntax
+
+Second, but still important - syntax.
+
+### Partiality Syntax
+
+Minimal Partiality we could write:
+```
+Partiality:      .{ PartialFields* }
+PartialFields:   PermittedField (, PermittedField )* ,?
+PermittedField:  IDENTIFIER | TUPLE_INDEX
+```
+
+### Partial Struct syntax
+
+Syntax is needed to Struct Type - is update `TypePath`
+```
+TypePath:   ::? TypePathSegment (:: TypePathSegment)* Partiality?
+```
+
+### Partial Tuple syntax
+
+For Tuple Type we need to update `TupleType`
+```
+TupleType:  ( ) | ( ( Type , )+ Type? ) Partiality?
+```
+
+### Partial Expression syntax
+
+For Expression we need create new kind of Expression:
+```
+PartialExpression:   Expression Partiality
+```
+
+and include it into `ExpressionWithoutBdeny`:
+```
+ExpressionWithoutBdeny:   ... | FieldExpression | PartialExpression | ...
+```
+
+
+## Logic Scheme
+
+Third, but still important - Logic Scheme.
+
+For pseudo-rust we suppose, partiality is a `HashSet` of permitted field-names.
+
+Common rules:
+```rust
+fn bar(v : SomeType.{'type_prtlty}) 
+{ /* .. */ }
+
+let v : SomeType.{'var_prtlty}; 
+```
+Then:
+
+(1) If `SomeType` is not supported type (neither Struct nor Tuple) then Error.
+
+(2) If partiality has no extra field-names `type_prtlty.is_subset(full_prtlty)` it compiles, otherwise Error.
+
+(3) If `var_prtlty.is_subset(full_prtlty)` it compiles, otherwise Error.
+
+(4) If `type_prtlty.is_empty()` or `var_prtlty.is_empty()` (if they are explicitly written as '`.{}`') then Error
```

Comment:
it could be reasonable to allow empty sets, if the function only used the address for comparison and wasn't allowed to read/write any fields.